Examples/Io/Root/src/RootNTupleReader.cpp

```diff
@@ -0,0 +1,272 @@
+  m_inputChain->SetBranchAddress("track_cov_phitheta",
+                                 &m_branches.m_track_cov_phitheta);
+  m_inputChain->SetBranchAddress("track_cov_phiqOverP",
+                                 &m_branches.m_track_cov_phiqOverP);
+  m_inputChain->SetBranchAddress("track_cov_tehtaqOverP",
+                                 &m_branches.m_track_cov_tehtaqOverP);
+
+  // Truth vertex
+  m_inputChain->SetBranchAddress("truthvertex_x", &m_branches.m_truthvertex_x);
+  m_inputChain->SetBranchAddress("truthvertex_y", &m_branches.m_truthvertex_y);
+  m_inputChain->SetBranchAddress("truthvertex_z", &m_branches.m_truthvertex_z);
+  m_inputChain->SetBranchAddress("truthvertex_t", &m_branches.m_truthvertex_t);
+
+  m_inputChain->SetBranchAddress("recovertex_x", &m_branches.m_recovertex_x);
+  m_inputChain->SetBranchAddress("recovertex_y", &m_branches.m_recovertex_y);
+  m_inputChain->SetBranchAddress("recovertex_z", &m_branches.m_recovertex_z);
+  m_inputChain->SetBranchAddress("truthvertex_tracks_idx",
+                                 &m_branches.m_truthvertex_tracks_idx);
+
+  m_inputChain->SetBranchAddress("beamspot_x", &m_branches.m_beamspot_x);
+  m_inputChain->SetBranchAddress("beamspot_y", &m_branches.m_beamspot_y);
+  m_inputChain->SetBranchAddress("beamspot_z", &m_branches.m_beamspot_z);
+  m_inputChain->SetBranchAddress("beamspot_sigX", &m_branches.m_beamspot_sigX);
+  m_inputChain->SetBranchAddress("beamspot_sigY", &m_branches.m_beamspot_sigY);
+  m_inputChain->SetBranchAddress("beamspot_sigZ", &m_branches.m_beamspot_sigZ);
+
+  auto path = m_cfg.inputFilePath;
+
+  // add file to the input chain
+  m_inputChain->Add(path.c_str());
+  ACTS_DEBUG("Adding File " << path << " to tree '" << m_cfg.inputTreeName
+                            << "'.");
+
+  m_events = m_inputChain->GetEntries();
+  ACTS_DEBUG("The full chain has " << m_events << " entries.");
+
+  {
+    // The entry numbers for accessing events in increased order (there could be
+    // multiple entries corresponding to one event number)
+    std::vector<long long> m_entryNumbers;
+    // If the events are not in order, get the entry numbers for ordered events
+    m_entryNumbers.resize(m_events);
+    m_inputChain->Draw("EventNumber", "", "goff");
+    // Sort to get the entry numbers of the ordered events
+    TMath::Sort(m_inputChain->GetEntries(), m_inputChain->GetV1(),
+                m_entryNumbers.data(), false);
+  }
+}
+
+std::pair<std::size_t, std::size_t>
+ActsExamples::RootNTupleReader::availableEvents() const {
+  return {0u, m_events};
+}
+
+ActsExamples::ProcessCode ActsExamples::RootNTupleReader::read(
+    const ActsExamples::AlgorithmContext& context) {
+  ACTS_DEBUG("Trying to read track parameters from ntuple.");
+
+  Acts::Vector3 pos(0, 0, 0);
+  std::shared_ptr<Acts::PerigeeSurface> surface =
+      Acts::Surface::makeShared<Acts::PerigeeSurface>(pos);
+
+  if (context.eventNumber >= m_events) {
+    ACTS_ERROR("event out of bounds");
```

Comment:
One last point: could we add a `return ActsExamples::ProcessCode::ABORT` here? (or whatever the error exit code is?)